pages/chaoss/visualizations/contrib_importance_over_time.py

```diff
@@ -0,0 +1,443 @@
+    while True:
+        period_from.append((end_date - window_width).strftime('%Y-%m-%d'))
+        period_to.append(end_date.strftime('%Y-%m-%d'))
+
+        # index df such that its rows are between the end date - window width and end date
+        mask = (df['created_at'] >= end_date - window_width) & (df['created_at'] <= end_date)
+        count_cntrbs_window = count_cntrbs(df.loc[mask], action_type, patterns)
+
+        # calculate contributor importance
+        cntrb_importance, cntrb_id, num_outliers = calc_cntrb_importance(count_cntrbs_window, action_type, threshold)
+        
+        cis.append(cntrb_importance)
+        
+        # shift the end_date by the step size
+        end_date -= step_size
+        
+        # get the contributor importance for the remaining slice, even if it is smaller than the window width
+        # set flag to indicate if the remaining slice has not been calculated yet
+        remaining_slice = False
+        if end_date - window_width <= start_date:
+            # calculate the remaining time between the current end date and start date, which will be the new window width
+            last_window_width = end_date - start_date
+            period_from.append((end_date - last_window_width).strftime('%Y-%m-%d'))
+            period_to.append(end_date.strftime('%Y-%m-%d'))
+
+            # index df such that its rows are betweeen the end_date - window_width and end date
+            remaining_mask = (df['created_at'] >= end_date - last_window_width) & (df['created_at'] <= end_date) 
+            count_cntrbs_last_window = count_cntrbs(df.loc[remaining_mask], action_type, patterns)
+            # calculate contributor importance for the remaining slice
+            cntrb_importance, cntrb_id, num_outliers = calc_cntrb_importance(count_cntrbs_last_window, action_type, threshold)
+            cis.append(cntrb_importance)
+            # set flag to True, indicating that the contributor importance for the remaining slice has been calculated
+            remaining_slice = True
+
+        # exit loop if the remaining slice has been calculated
+        if remaining_slice:
+            break
+        
+    # create a dataframe with period to, period from, contributor importance, and cntrb_ids as columns
+    df = pd.DataFrame(list(zip(period_from, period_to, cis)), columns = ['period_from', 'period_to', 'cntrb_importance'])
+    return df
+ 
+def create_figure(df: pd.DataFrame, action_type, step_size):
+    # create plotly express line graph
+    fig = px.line(df, x="period_from", y="cntrb_importance", custom_data=["period_to"], text="cntrb_importance")
+    
+    # format x-axis ticks to be in Month Year format
+    fig.update_xaxes(fixedrange=True, tickformat="%B %Y", tickangle=90, dtick=f"M{step_size}")
+
+    # define x-axis and y-axis titles and intialize first x-axis tick to start at the user-inputted start_date
+    start_date = min(df['period_from'])
+    fig.update_layout(
```

Comment:
Change to quarterly ticks (every 3 months), common in business vernacular. Also use shorted month names to take up less space. First Time Contributions Per Quarter graph is a good example of this

Reply:
Since the x-axes values are cyclical according to the step size, I spaced them out such that there's a tick for every step_size * 2 months.

Reply:
turn x axis to be parallel to the graph, allows more space for the graph